server/hooks.go

```diff
@@ -0,0 +1,68 @@
+package main
+
+import (
+	"fmt"
+
+	"github.com/mattermost/mattermost-server/v5/mlog"
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/mattermost/mattermost-server/v5/plugin"
+)
+
+// UserHasJoinedTeam is invoked after the membership has been committed to the database. If
+// actor is not nil, the user was added to the team by the actor.
+func (p *Plugin) UserHasJoinedTeam(c *plugin.Context, teamMember *model.TeamMember, actor *model.User) {
+	data := p.constructMessageTemplate(teamMember.UserId, teamMember.TeamId)
+	if data == nil {
+		return
+	}
+
+	for _, message := range p.getWelcomeMessages() {
+		if message.TeamName == data.Team.Name {
+			go p.processWelcomeMessage(*data, *message)
+		}
+	}
+}
+
+// UserHasJoinedChannel is invoked after the membership has been committed to
+// the database. If actor is not nil, the user was invited to the channel by
+// the actor.
+func (p *Plugin) UserHasJoinedChannel(c *plugin.Context, channelMember *model.ChannelMember, _ *model.User) {
+	if channelInfo, appErr := p.API.GetChannel(channelMember.ChannelId); appErr != nil {
+		mlog.Error(
+			"error occured while checking the type of the chanel",
+			mlog.String("channelId", channelMember.ChannelId),
+			mlog.Err(appErr),
+		)
+		return
+	} else if channelInfo.Type == model.CHANNEL_PRIVATE {
+		return
+	}
+
+	key := fmt.Sprintf("%s%s", channelMember.ChannelId, WELCOMEBOT_CHANNEL_WELCOME_KEY)
+	data, appErr := p.API.KVGet(key)
+	if appErr != nil {
+		mlog.Error(
+			"error occured while retrieving the welcome message",
+			mlog.String("channelId", channelMember.ChannelId),
+			mlog.Err(appErr),
+		)
+		return
+	}
+
+	if data == nil {
+		// No welcome message for the given channel
+		return
+	}
+
+	post := &model.Post{
+		UserId:    p.botUserID,
+		ChannelId: channelMember.ChannelId,
+		Message:   string(data),
+	}
+	if _, appErr := p.API.CreatePost(post); appErr != nil {
+		mlog.Error("failed to post welcome message to the channel",
+			mlog.String("channelId", channelMember.ChannelId),
+			mlog.Err(appErr),
+		)
+	}
```

Comment:
I think `p.API.SendEphemeralPost` doesn't work here, because the (webapp) client doesn't know about the newly joined channel yet. Hence, it's a race between `SendEphemeralPost` and the information about the newly created channel. If the web socket event from `SendEphemeralPost` is received first, the client just drops it and no ephemeral message is shown.


Something like `time.Sleep()` helps a bit, but in a mobile setup this might not even help.

There is another issue with using an ephemeral message: If the user is not online while the invite happens, he/she will not get the message.

I'm wondering if a DM from the bot should be used instead. This has the advantage that other users don't get spammed and no ephemeral message is used. On the other side it means the welcome post is not longer in the same context as the newly joined channel.

This discussion does not have to be resolved bevor merging this PR, but I would like to hear some thoughts first and get a general direction.

Reply:
Imagine sending a welcome message via a bot DM when the user is invited to multiple channels in short succession - would it be readable? Also, it may be a matter of taste, but at least for me, it would no longer sound like a channel welcome message but more like a bot welcome message.

I do not know how difficult would be to fix the race condition described and/or if this is by design, but this sounds like a problem with mm-server/webapp that should be addressed. 

One could imagine a hybrid approach - fix the race, and send ephemeral messages if the user is online (can we check it from within plugin?), or send a bot DM otherwise with some fancy header ("Hey, I noticed that you were invited to X, but you were unavailable. See the welcome message below:..."). PM would have to chime in as well I think.

I am a fan of splitting work into smaller chunks and iterating, even though what we have is not perfect. Hence my vote would be to have two followup issues and for now, just send plain posts to the channel.

2cents apply

Reply:
I've many thoughts on this issue. 

1. I don't think we should ship with the welcomeBot non-ephemeral posts in channels.  Others should not see the message everytime someone joins a channel. They already see who is added to a channel.
2. I actually favor the WelcomeBot DM for the following reasons:
  1. Keeps a record of the message received from welcomeBot when joining a channel (instead of ephemeral post). 
  2. welcomeBot DM is currently under-utilized, in my opinion.  I have only 3 messages from welcomeBot over my first year and none since before the day I joined MM.
  3. If the welcome message doesn't provide any additional and important information that is worth keeping around for later reference, do I really care to get a UX hello that disappears.
  4. When you are added to channel, the channel sidebar shows the channel with an unread message. At a quick glance, it looks like any other channel with mentions.  If you received a DM from welcomeBot, you'd know that you got welcomed to a channel that didn't exists.  

That's a lot of thoughts with my personal preference, but I'm not the authority to make the decision.

Reply:
@aaronrothschild did you have a chance to weigh in on ^^?

My own $.02: do what @vespian proposed: a DM, and an optimistic attempt to post an ephemeral message, `time.Sleep(1*time.Second)` seems fine to me. Agree that we should not (normally?) be posting a welcome to the channel, the users are already notified when a new member joins, right?